tools/neighbrwatch/neighbrwatch.man
.TH OLED NEIGHBRWATCH 8 "Aug 2025" "Oracle Linux Enhanced Diagnostics"

.SH NAME
oled neighbrwatch - Read and process ip-neighbor entries

.SH SYNOPSIS
oled neighbrwatch [-h] [-i INFILES [INFILES ...]] [-t] [-a ADDR] [-m MAC]
                  [-c COUNT] [-T [INTERVAL]] [-C [MAX_SAMPLE]]
                  [-o OUTFILE]


.SH DESCRIPTION

neighbrwatch tool will show all mac addresses mapped to each IP address. It can read and process neighbor entries either dynamically fetched from the current server or stored in a log collected from RDSinfo exawatcher script. The tool can show discrepancies in mac assignment (could help in duplicate IP issue) if an IP address shows MAC association more than others.  It accepts -T (interval [def: 2sec]) and -C (max-sample [def:5]) and reads neighbor entries every -T seconds for -C times.
At the end, it displays list of IP address and its linked MAC addresses.

The output can be filtered based on IP_address (-a IP) or mac address (-m mac). If input file is given, tool will only process from the file. To read neighbor entries dynamically, then do not give input file.

When file is given, the option -t will show entry count in each samples in the file. If input file is given, tool will only process from the file. To read neighbor entries dynamically, then do not give input file.

If no argument is given, the tool reads dynamic data based on default values of interval and max-samples
  

.SH OPTIONS
.TP
  -h, --help
	Displays a help message for neighbrwatch command.

.TP
  -i INFILES [INFILES ...], --infiles INFILES [INFILES ...]
        Files from RDSinfo exawatcher logs. To process already captured data.

.TP
  -t
        Count entries per sample in files. Use with -i option.

.TP
  -a ADDR, --addr ADDR
        Display MACs matching given IP address

.TP
  -m MAC, --mac MAC
        Display IP addresses matching given MAC

.TP
  -c COUNT, --count COUNT
        Display entries matching >= 'count' MACs for an IP address

.TP
  -T [INTERVAL], --interval [INTERVAL]
        Collect data every T seconds [default=2]

.TP
  -C [MAX_SAMPLE], --max_sample [MAX_SAMPLE]
        Collect C samples [default=5]

.TP
  -o OUTFILE, --outfile OUTFILE
        Write output to a file

.SH EXAMPLES
# oled neighbrwatch
.br
Reading neighbor table: interval=2 count=5
.br
------------------------------------------------------------
           IP_addr             cnt                  mac
.br
------------------------------------------------------------
  1) 10.211.8.1                  1   [ 00:00:0c:9f:f0:1f ]
  2) 10.211.8.19                 1   [ a4:4c:11:6c:c8:3f ]
  3) 10.211.8.2                  1   [ 58:ac:78:e3:66:7f ]
  4) 10.211.8.3                  1   [ 58:ac:78:e3:59:ff ]
  5) 10.211.8.40                 1   [ 56:6f:f5:fe:00:a9 ]
  6) 10.211.9.162                1   [ 00:10:e0:e7:43:52 ]
.br
------------------------------------------------------------

# oled neighbrwatch -i /tmp/RDSInfo-exawatcher*.dat
.br
Reading neighbor from files
.br
------------------------------------------------------------
           IP_addr             cnt                  mac
.br
------------------------------------------------------------

  1) 192.168.76.10               1   [ 52:54:00:a4:ae:aa ]
  2) 192.168.76.50               2   [ 52:54:00:a4:ae:cc, 52:54:00:a4:ae:0b ]
  3) 192.168.76.55               1   [ 52:54:00:a4:ae:11 ]
  4) 192.168.76.cc               1   [ 52:54:00:a4:ae:cc ]
.br
------------------------------------------------------------

# oled neighbrwatch -i /tmp/ntable*.dat -t
.br
Reading neighbor from files
.br
------------------------------------------------------------

  1) file: /tmp/ntable1.dat

zzz <07/28/2025 13:53:14> subcount: 0
.br
Number of entries = 37

zzz <07/28/2025 13:54:14> subcount: 1
.br
Number of entries = 32

zzz <07/28/2025 13:55:15> subcount: 2
.br
Number of entries = 32

zzz <07/28/2025 13:56:16> subcount: 3
.br
Number of entries = 32


  2) file: /tmp/ntable2.dat

zzz <07/29/2025 10:53:14> subcount: 0
.br
Number of entries = 10

zzz <07/29/2025 10:54:14> subcount: 1
.br
Number of entries = 32

zzz <07/29/2025 10:55:15> subcount: 2
.br
Number of entries = 32

zzz <07/29/2025 10:56:16> subcount: 3
.br
Number of entries = 32


# oled neighbrwatch -i /tmp/RDSInfo-exawatcher*.dat -c 2
.br
Reading neighbor entry from files
.br
------------------------------------------------------------
.br
           IP_addr             cnt                  mac
.br
------------------------------------------------------------
.br
  1) 192.168.76.50               2   [ 52:54:00:a4:ae:cc, 52:54:00:a4:ae:0b ]
.br
------------------------------------------------------------


.SH REPORTING BUGS
.TP
	Please contact Oracle Linux Support to report any bugs for OLED tools.

.SH AUTHOR
.TP
        Arumugam Kolappan <aru.kolappan@oracle.com>


tools/neighbrwatch/neighbrwatch.py
#!/usr/bin/env python3
#
# Copyright (c) 2025, Oracle and/or its affiliates.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This code is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 only, as
# published by the Free Software Foundation.
#
# This code is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# version 2 for more details (a copy is included in the LICENSE file that
# accompanied this code).
#
# You should have received a copy of the GNU General Public License version
# 2 along with this work; if not, see <https://www.gnu.org/licenses/>.
#
# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
# or visit www.oracle.com if you need additional information or have any
# questions.

# Author: Arumugam Kolappan <aru.kolappan@oracle.com>

"""
This tool displays IP neighbor entries reading from files or dynamically
fetch from the server. The input files are collected from RDSInfo exawatcher
logs. The script can display IP-MAC mapping based on either an IP address
or MAC address, show number of neighbor entries per sample, display
IP/mac based on MAC count etc.
"""

import sys
import time
import argparse
import subprocess

MAC_LIST = {}


def store_neighbor(ip_addr, mac):
    """
    Add mac address to the list
    """
    if ip_addr not in MAC_LIST:
        MAC_LIST[ip_addr] = set()
    MAC_LIST[ip_addr].add(mac)


def read_neighbor(files, fpopen):
    """
    Process mac from files
    """
    print(f"Reading neighbor from files...", file=fpopen)
    for infile in files:
        try:
            with open(infile, "r") as flog:
                for line in flog.readlines():
                    words = line.split()
                    if words and len(words) > 5:
                        if words[1] == "dev" and words[3] == "lladdr":
                            ip_addr = words[0]
                            mac = str(words[4])
                            store_neighbor(ip_addr, mac)
        except (PermissionError, FileNotFoundError, IOError,
                IsADirectoryError, UnicodeDecodeError):
            print("Error: File error: {0}".format(infile))
            sys.exit(1)


def show_neighbor(cond, args, fpopen):
    """
    Print the IP+mac collected
    """
    print('-' * 60, file=fpopen)
    print("{:>18}{:>16}{:>21} ".format("IP_addr", "cnt", "mac"), file=fpopen)
    print('-' * 60, file=fpopen)
    matching = False

    cnt = 1
    for ip_addr in sorted(MAC_LIST):
        mac = MAC_LIST[ip_addr]
        if cond == "both" and ip_addr == args.addr and args.mac in mac:
            matching = True
        elif cond == "addr" and ip_addr == args.addr:
            matching = True
        elif cond == "mac" and args.mac in mac:
            matching = True
        elif cond == "count" and len(mac) >= int(args.count):
            matching = True
        elif cond == "list":
            matching = True

        if matching:
            print("{:>3}) {:<25}   {:<3} [ {} ]".format(cnt, ip_addr, len(mac),
                                                        ", ".join(mac)),
                  file=fpopen)
            cnt += 1
            matching = False
    print('-' * 60, file=fpopen)


def process_one_entry(line, prev_cnt, fpopen) -> int:
    """
    process single neighbor entry
    """
    total = prev_cnt
    words = line.split()
    if words:
        if words[0] == "zzz" and words[3] == "subcount:":
            if total > -1:
                print("Number of entries = {}\n".format(total), file=fpopen)
            total = 0
            print(line, file=fpopen)
            return total

        if (len(words) > 5 and words[1] == "dev" and words[3] == "lladdr"):
            total = total + 1
    return total


def show_neigh_count(files, fp_log):
    """
    Print number of entries in each sample
    """
    cnt = -1
    if files:
        print('-' * 60, file=fp_log)
        sno = 1
        for infile in files:
            try:
                with open(infile, "r") as fp_in:
                    print(f"\n{sno:>3}) file: {infile}\n", file=fp_log)
                    sno += 1
                    cnt = -1
                    for line in fp_in.readlines():
                        line = line.strip()
                        cnt = process_one_entry(line, cnt, fp_log)
                    if cnt > -1:
                        print("Number of Entries = {}\n".format(cnt),
                              file=fp_log)
                    else:
                        print("Number of Entries = 0\n", file=fp_log)
            except (PermissionError, FileNotFoundError, IOError):
                print("Error: File error: {0}".format(infile))
                return


def open_logfile(fname):
    """
    Open the file to read
    """
    try:
        if fname:
            fopen = open(fname, "w")
        else:
            fopen = sys.stdout
    except (PermissionError, FileNotFoundError, IOError):
        print("Error: File error: {0}".format(fname))
        fopen = sys.stdout
        sys.exit(1)
    return fopen


def collect_one_neigh(line):
    """
    Process one neighbor entry
    """
    words = line.split()
    if words and len(words) > 5:
        if (words[1] == "dev" and words[3] == "lladdr"):
            ip_addr = words[0]
            mac = str(words[4])
            store_neighbor(ip_addr, mac)


def process_neighbor(interval, count, fpopen):
    """
    Process neigh entry from the system
    """
    print(f"Reading neighbor from server: interval={interval} count={count}",
          file=fpopen)
    i = 0
    while i < int(count):
        try:
            data = subprocess.run("/usr/sbin/ip neigh", shell=True, check=True,
                                  stdout=subprocess.PIPE,
                                  stderr=subprocess.PIPE)
        except subprocess.CalledProcessError:
            print(f"command failed: err: {data.returncode}. (iteration: {i})")
            return

        for line in data.stdout.splitlines():
            collect_one_neigh(line.decode('utf-8'))
        time.sleep(interval)
        i = i+1


def get_args() -> argparse.Namespace:
    """
    Read the arguments
    """

    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--infiles", nargs="+",
                        help="Files from RDSinfo exawatcher logs. "
                        "To process already captured data.")
    parser.add_argument("-t", action='store_true',
                        help="Entry count per sample in files. "
                        "Use with -i option")
    parser.add_argument("-a", "--addr",
                        help="Display MACs matching given IP address")
    parser.add_argument("-m", "--mac",
                        help="Display IP addresses matching given MAC")
    parser.add_argument("-c", "--count",
                        help="Display entries matching >= 'count' "
                        "MACs for an IP address")
    parser.add_argument("-T", "--interval",
                        help="Collect data every T seconds [default=2]",
                        nargs='?', type=int, const=2, default=2)
    parser.add_argument("-C", "--max_sample",
                        help="Collect C samples [default=5]", nargs='?',
                        type=int, const=5, default=5)
    parser.add_argument("-o", "--outfile", help="Write output to a file")
    args = parser.parse_args()
    return args


def main():
    """
    neighbrwatch will show all mac addresses mapped to each IP address. This
    can show discrepancies in mac assignment (could help in duplicate IP issue)
    if an IP address shows MAC association more than others. It processes
    entries from the log file as well as read current data from running system.
    Moreover, it provides other options such as: show IP-MAC list based on IP
    or mac address, neighbor entry count in each section in the log file etc.
    The output can be written to a file.
    """

    args = get_args()

    fp_open = open_logfile(args.outfile)

    if args.infiles:
        # read from file
        if args.outfile:
            print(f"Reading neighbor from files...")
        read_neighbor(args.infiles, fp_open)
    else:
        # read from system
        if args.outfile:
            print(f"Reading neighbor from server: interval={args.interval} ",
                  f"count={args.max_sample}")
        process_neighbor(args.interval, args.max_sample, fp_open)

    if args.t and args.infiles:
        show_neigh_count(args.infiles, fp_open)
    elif args.addr and args.mac:
        show_neighbor("both", args, fp_open)
    elif args.addr:
        show_neighbor("addr", args, fp_open)
    elif args.mac:
        show_neighbor("mac", args, fp_open)
    elif args.count:
        show_neighbor("count", args, fp_open)
    else:
        show_neighbor("list", 0, fp_open)

    fp_open.close()


if __name__ == "__main__":
    main()
